### PR TITLE
dockerfile first draft

### PR DESCRIPTION
**Pull request by choldgraf**
hey @yuvipanda , here's a first take at converting your notes to a proper doc page. It's still WIP but I'd appreciate your thoughts on it thus far. There are a few places in the notes where I need to fill in some info, and I'm also still not positive that this repository is the right place for this, but we can discuss :-)

http://predictablynoisy.com/binder/

TODO:

- [x] Add a section on reproducibility with Dockerfiles

### CODE REVIEW

**Comment by minrk** [comment]
It would be great to add at the top, something pointing to "If you find you have to use a Dockerfile, please let us know why. We hope to get to a point where nobody ever has to resort to a custom Dockerfile."

**Comment by choldgraf** [comment]
@minrk can you write down some quick points to cover re: reproducibility with dockerfiles? I'll flesh it out and add to this PR

**Comment by minrk** [comment]
Points re: reproducibility with Dockerfiles:

1. Always tag your base image! Use `FROM base/image:tag` instead of `FROM base/image`. The default (`:latest`) tag is ~guaranteed to change in breaking ways as upstream software gets updates. What tag to use depends on the tagging scheme of your base image. Docker tags are like git branches and can change. The safest choice is to use docker sha tags, e.g.

       FROM ubuntu@sha256:84c334414...

   which are more like git commits in that they are guaranteed to refer to a single version of the image.

2. Binder's requirements for Dockerfiles are in beta and changing, so Dockerfiles on Binder **will break** from time to time until Binder solidifies its requirements. Dockerfiles are not an appropriate choice for reproducibility with Binder at the moment.

And re: reproducibility in general:

1. Pin your dependencies when installing. When you install a dependency, don't just specify `numpy`, specify `numpy==1.12.0`. `pip freeze` is a handy tool to export the exact version of every Python package in your environment in a format that can be used in `requirements.txt`. `conda env export -n <env-name>` is the equivalent for conda `environment.yml`.

**Comment by choldgraf** [comment]
ok, added the language that @minrk suggested and cleaned things up a bit...think we can merge this? It definitely still has WIP components to it, but I think it does have valuable info

**Comment by minrk** [comment]
👍 @willingc do you want to give it a pass and then merge?

File: doc/dockerfile.rst [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    # Note that there must be a tag
+    FROM jupyter/scipy-notebook:cf6258237ff9
+
+See `Preparing your Dockerfile`_ for instructions on how to
+do this properly.
+
+When you are building complex software
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Some software is simply too complex to be specified with Binder configuration
```
Comment (willingc) [inline doc/dockerfile.rst]: Most Binder configurations can be achieved without a Dockerfile. Before resorting to a Dockerfile, we recommend trying to use ``postBuild`` commands for configuration.

File: doc/dockerfile.rst [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+Dockerfile?
+
+When you're using a language that is not directly supported
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Binder supports many languages, but not all of them. If your need to use
+a different language, it may be possible to accomplish this with a Dockerfile.
+
+For a list of languages that Binder supports with configuration files, see
+``the repo2docker documentation <http://repo2docker.readthedocs.io/en/latest/>`_.
```
Comment (willingc) [inline doc/dockerfile.rst]: ` Extra backtick.

File: doc/dockerfile.rst [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+   We welcome contributions to ``repo2docker`` to add support for new
+   languages. If interested, please
+   `open an issue <https://github.com/jupyter/repo2docker/issues>`_.
+
+
+Preparing your Dockerfile
+=========================
+
+For a Dockerfile to work on binder, it must meet the following requirements:
```
Comment (willingc) [inline doc/dockerfile.rst]: Binder

File: doc/dockerfile.rst [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+          RUN pip3 install --no-cache-dir jupyterhub=${JUPYTERHUB_VERSION}
+
+2. It must explicitly specify a tag in the image you source.
+
+   When sourcing a pre-existing Docker image with ``FROM``,
+   **a tag is required**. The tag *cannot* be ``latest``. Note that tag
+   naming conventions differ between images, so we recommend using
+   the SHA tag of the image.
+
+   Here's an example of a Dockerfile ``FROM`` statement that would work:
```
Comment (willingc) [inline doc/dockerfile.rst]: ... work.

**Comment by choldgraf** [comment]
ok @willingc updates made, LMK

File: doc/index.rst [hunk 5]
```diff
@@ -95,3 +95,20 @@ package in your environment in a format that can be used in ``requirements.txt``
 
 ``conda env export -n <env-name>`` is the equivalent for anaconda's environment.yml
 file.
+
+If using a Dockerfile, source a reproducible, tagged image
```
Comment (willingc) [inline doc/index.rst]: Let's move all of this over to the `dockerfile.rst` document. My understanding is that we are trying to guide folks to using Dockerfiles as a choice in a very limited use case (i.e. when pip and conda can't get you what is needed). I think it's fine to have a short paragraph or a few bullets here, but the current text is mixing implementation with general tips for reproducibility.

Comment (choldgraf) [reply doc/index.rst]: +1

**Comment by choldgraf** [comment]
updated per @willingc 's comments...LMK if you'd like me to change some other language in there. I think we should probably have another PR to flesh out the reproducibility tips more in the future (maybe when we create the reproducibility validator machinery)

File: doc/index.rst [hunk 6]
```diff
@@ -96,19 +96,9 @@ package in your environment in a format that can be used in ``requirements.txt``
...
-image. This ensures that you will continue building off of the same image
-even if the image is updated to a new version.
-
-Second, make sure that all packages installed with your Dockerfile
-are pinned to specific versions (you should do this with the the image you are
-sourcing as well).
+Using Dockerfiles
+`````````````````
 
+Ensuring reproducibility with Dockerfiles comes with its own set of challenges.
```
Comment (willingc) [inline doc/index.rst]: Love this wording.

**Comment by willingc** [comment]
Thanks @choldgraf for the edits. It looks good.

**Comment by choldgraf** [comment]
Yay!

**Pull request merged**